### ISSUE TITLE
Evaluation dataset (result.h5) and calculation of Error

### ISSUE THREAD

**Issue by Aqsa-sk** [deepqmc/deepqmc#229]
I am little puzzled about the size of dataset generated during training and evaluation. Following are the examples of h5ls outputs for various electron_batch_size  during evaluation on the same trained checkpoint. Here, I provide steps and batch size for training data, then evaluate its first checkpoint using different batch sizes. For evaluation same train() function was used only the optimiser was set as none as follows:

```
step, train_state = CheckpointStore.load(f'../runs/{path}/training/chkpt-{checkpoint}.pt'

train(
    H, ansatz,
    None,                              # disable optimizer, for traininig it was kfac
    sampler_factory,
    train_state=train_state,
    steps=n_steps,                     # number of sampling steps for evaluation
    electron_batch_size=e_batch_size,
    seed=n_seed,
    workdir=f'run/'  # folder for evaluation logs
   #chkpt_constructor = partial(CheckpointStore, interval=5) # no chkpt for evaluation
   )
```

Here are h5ls output for training and evaluation steps: (I am using 4 gpus for each run)

**training:**
steps = 10
electron_batch_size = 40
```
$ h5ls -r result.h5
/                        Group
/local_energy            Group
/local_energy/max        Dataset {10/Inf, 1, 1}
/local_energy/mean       Dataset {10/Inf, 1, 1}
/local_energy/min        Dataset {10/Inf, 1, 1}
/local_energy/samples    Dataset {10/Inf, 4, 1, 1, 10}
/local_energy/std        Dataset {10/Inf, 1, 1}
/psi                     Group
/psi/samples             Group
/psi/samples/log         Dataset {10/Inf, 4, 1, 1, 10}
/psi/samples/sign        Dataset {10/Inf, 4, 1, 1, 10}
/sampling                Group
/sampling/log_psi        Group
/sampling/log_psi/mean   Dataset {10/Inf, 1, 1}
/sampling/log_psi/std    Dataset {10/Inf, 1, 1}
```

**evaluation:**
steps = 10
electron_batch_size = 1600
```
$ h5ls -r result.h5
/                        Group
/local_energy            Group
/local_energy/max        Dataset {10/Inf, 1, 1}
/local_energy/mean       Dataset {10/Inf, 1, 1}
/local_energy/min        Dataset {10/Inf, 1, 1}
/local_energy/samples    Dataset {10/Inf, 4, 1, 1, 10} ------------> same shape as training even though batch size is 1600
/local_energy/std        Dataset {10/Inf, 1, 1}
/psi                     Group
/psi/samples             Group
/psi/samples/log         Dataset {10/Inf, 4, 1, 1, 10}
/psi/samples/sign        Dataset {10/Inf, 4, 1, 1, 10}
/sampling                Group
/sampling/log_psi        Group
/sampling/log_psi/mean   Dataset {10/Inf, 1, 1}
/sampling/log_psi/std    Dataset {10/Inf, 1, 1}
```

**evaluation:**
steps = 10
electron_batch_size = 16

```
/                        Group
/local_energy            Group
/local_energy/max        Dataset {10/Inf, 1, 1}
/local_energy/mean       Dataset {10/Inf, 1, 1}
/local_energy/min        Dataset {10/Inf, 1, 1}
/local_energy/samples    Dataset {10/Inf, 4, 1, 1, 10} ------------> same shape as training even though batch size is 16
/local_energy/std        Dataset {10/Inf, 1, 1}
/psi                     Group
/psi/samples             Group
/psi/samples/log         Dataset {10/Inf, 4, 1, 1, 10}
/psi/samples/sign        Dataset {10/Inf, 4, 1, 1, 10}
/sampling                Group
/sampling/log_psi        Group
/sampling/log_psi/mean   Dataset {10/Inf, 1, 1}
/sampling/log_psi/std    Dataset {10/Inf, 1, 1}
```

Does batch size provided during evaluation step not reflect in the data set for /local_energy/samples?

While comparing with CLI output for energy mean and error calculation, I am unable to reproduce same number using dataset /local_energy/mean and /local_energy/std  and /local_energy/samples.

For both training and evaluation I am using `sampler_factory = partial(initialize_sampling, elec_sampler=MetropolisSampler)`
As per another open [issue](https://github.com/deepqmc/deepqmc/issues/227), I am able to get close statistics to the CLI output by using,
E_mean = average(dataset /local_energy/mean) = -7.51817022 (whereas CLI prints -7.92(35)).
I am not sure how to get the error of (35) from either /local_energy/std or /local_energy/samples dataset. Are there any equilibrium steps that need to be accounted for or decorrelation effect?

Any help will be highly appreciated in order to reproduce CLI number (I can't always run CLI since my jobs are run through slurm).

**Comment by szbernat** [comment]
Hi Aqsa,

Thanks for your continued interest, we are really hoping that deepQMC will be a useful tool for you!

Zeno is at the moment composing detailed answers to your questions about computing statistical errors. I just wanted to quickly jump in to advertise using deepQMC via the command line interface.

Using the command line executable has many advantages over directly calling the `train` function from a python script:

1. It's reproducible: every run started from the command line automatically writes the settings it used in the file `workdir/training/.hydra/config.yaml`. This means that you don't have to remember or manually save the settings you used for all of your runs, you can always check them in these autogenerated files.
2. It's less verbose: the CLI has sensible defaults for most settings, so you only have to change those settings which you care about. It also takes care of some annoying tasks which you have to do yourself in a python script like loading the trained checkpoint when doing evaluation.
3. You can use the predefined config groups that abstract away a lot of the unimportant details: for example if you want to switch from the default Langevin sampler to Metropolis-Hastings, you don't have to specify all the details of the sampling algorithm, it's enough to use the predefined `decorr_metropolis.yaml` config file via passing `sampler_factory=decorr_metropolis` in the `deepqmc` command.

We ourselves use the CLI almost exclusively to run all of our experiments with `deepqmc`, i.e. we never write python scripts which explicitly call the `train` function. You can find a basic introduction to using the `deepqmc` command here: https://deepqmc.github.io/cli.html We are also happy to help you along the way if any questions come up.

**Comment by zenoone** [comment]
Thanks Bernat, I agree that using the CLI is the easier and more reproducible, hence preferred, way of running calculations.

Before we get to the issue with the dataset sizes lets briefly discuss the other questions. 

> Are there any equilibrium steps that need to be accounted for or decorrelation effect?

As answered in #228 there are no such things you need to take into account.

> While comparing with CLI output for energy mean and error calculation, I am unable to reproduce same number using dataset /local_energy/mean and /local_energy/std and /local_energy/samples.

As described in #227 computing the energy and energy error from the `results.h5` file is the correct way and the numbers from the progress bar are just a guidance throughout the training/evaluation. Btw. you can also compute the mean and error estimate from the `results.h5` files during training/evaluation if you use the single write multiple read option in h5py:
```with h5py.File(hdf5_file, 'r', swmr=True) as f: ...``` 

Regarding the sizes of the datasets. What you report is not the expected behavior and I cannot reproduce it on my side. The `electron_batch_size` of the evaluation should be present in the respective `results.h5` file. Could it be that you mixed up the paths to `results.h5`. Note that if you evaluate (not provide an optimizer) your path will be `.../evaluation/result.h5` instead of `.../training/result.h5`.

**Comment by Aqsa-sk** [comment]
Thank you for your detailed response. I appreciate it! :)
I am able to  correctly obtain the Error on energy using the suggestions provided in [issue](https://github.com/deepqmc/deepqmc/issues/228)

Regarding the sizes of data set. I have found something interesting, not sure if it is the way I am implementing or a bug.

So when I start the train from scratch as (eg for O atom),
```
mol = Molecule(  # O
    coords=[[0.0, 0.0, 0.0]],
    charges=[8],
    charge=0,
    spin=2,
    unit='angstrom',
)
H = MolecularHamiltonian(mol=mol, ecp_type='ccECP', ecp_mask=[True], elec_std=0.1)

deepqmc_dir = os.path.dirname(deepqmc.__file__)
config_dir = os.path.join(deepqmc_dir, 'conf/ansatz')
with initialize_config_dir(version_base=None, config_dir=config_dir):
    cfg = compose(config_name='default')
_ansatz = instantiate(cfg, _recursive_=True, _convert_='all')
ansatz = instantiate_ansatz(H, _ansatz)

elec_sampler = partial(combine_samplers, samplers=[DecorrSampler(length=20), MetropolisSampler])
sampler_factory = partial(initialize_sampling, elec_sampler=elec_sampler)

config_dir = os.path.join(deepqmc_dir, 'conf/task/opt')
with initialize_config_dir(version_base=None, config_dir=config_dir):
    cfg = compose(config_name='kfac')
kfac = instantiate(cfg, _recursive_=True, _convert_='all')

n_steps = 10
e_batch_size = 12
n_seed = 71
path = 8

train(H, ansatz, kfac, sampler_factory,
      steps = n_steps,
      electron_batch_size = e_batch_size,
      seed=n_seed, workdir = f'runs-12/{path}',
      chkpt_constructor = partial(CheckpointStore, interval=50)
      )

```

The dataset shapes are consistent with the batch size provided in the above python script. 

But, when I restart the train  from existing checkpoint (checkpoint path compared to the current python file: ../runs-12/8/training/chkpt-10.pt) with a new batch size, it does not respect the "new value" of batch size and provides me result.h5 with same shape ( `/local_energy/samples    Dataset {10/Inf, 4, 1, 1, 3}`) of training data set despite having different batch sizes. I am employing the following modifications to restart the training (everything else in the program remains the same just the train function now has train_state pointing to the checkpoint and a different batch size).

```

## Load the checkpoint
step, train_state = CheckpointStore.load(f'../runs-12/8/training/chkpt-10.pt')

n_steps = 10
e_batch_size = 20
n_seed = 68
path = 8

start = time.time()
train(H, ansatz, kfac, sampler_factory,
      train_state=train_state,
      steps = n_steps,
      electron_batch_size = e_batch_size,
      seed=n_seed, workdir = f'runs/{path}',
      chkpt_constructor = partial(CheckpointStore, interval=10)
      )

```

Similar situation arises for evaluation as well, where I write None at the place of kfac (since no more optimization needed) and change the batch size and/or seed while using train state as the last trained checkpoint.

**Looks like whenever I am working on existing checkpoint of a certain batch size my future runs on it follow the same batch size regardless of what I am providing in the train function.**

For reproducibility I am attaching the complete python scripts. Steps to run them:
1. Download and rename .txt extension to .py
2. run first python in path/O.py
3. mkdir continue
4. run second python script in path/continue/O-restart.py

[O.txt](https://github.com/user-attachments/files/21673911/O.txt)
[O-restart.txt](https://github.com/user-attachments/files/21673912/O-restart.txt)


On a different note, I like your suggestion of using CLI rather than python but I am more comfortable with python than hydra and also use slurm scripts for my job. I haven't tried the same change of batch sizes while using checkpoints in CLI since I am unfamiliar with hydra. It would be good to cross check with CLI implementation.

Can you suggest me how to efficiently use CLI command to train retrain and evaluate multiple different system for various ansatz and steps, batch sizes (as an example how to do the same attached O.py and O-restart.py using CLI)? Do I write my own yaml file in the source directory for certain systems and use hydra to overwrite parameters when needed changed? And for slurm  can I run the same command as a bash file in my job script?

Thank you for your detailed responses. They have been very helpful.

**Comment by zenoone** [comment]
Thanks for providing the scripts, I see the problem now. What you observe is indeed expected behavior. However, there is an important detail about continued training and evaluation, that I realize is not well documented on in the manual. When loading a checkpoint the [TrainState](https://deepqmc.github.io/api.html#training-and-evaluation) contains the model parameters, alongside the sampler state and the optimizer state. If you directly pass that TrainState to the train function you will exactly restart from the checkpoint, that is all the existing Markov chains will be continued instead of initializing new walkers, and hence the electron_batch_size argument is ignored. If instead you want to continue with a different batch size, you need to remove the sampler state. Add this line after loading the checkpoint:
```
# Load the checkpoint
step, train_state = CheckpointStore.load(f'../runs-12/8/training/chkpt-10.pt')
# Remove sampler state
train_state = train_state._replace(sampler=None)
```
Now your new electron batch size will be respected. Sorry for the inconvenience. 
Here is a basic example of how you could use DeepQMC through the cli:
To run `O.py` via the command line, the command would be:
```
deepqmc task.steps=10 task.seed=8 hydra.run.dir='runs-12/9' task/sampler_factory=decorr_metropolis task.electron_batch_size=12 +task.chkpt_constructor.interval=5
```
To continue a job that somehow got cannceled during the training process you would use:
```
deepqmc task=restart hydra.run.dir='runs-12/9_restart' task.restdir='runs-12/9/training'
```
If you want to continue from a checkpoint, but change arguments similar to what you did in `O-restart.py` you can run:
```
deepqmc task=restart hydra.run.dir='runs-12/9_continued' task.restdir='runs-12/9/training' +task.electron_batch_size=20 +task.steps=20 +task.chkpt='chkpt-10.pt' +task.keep_sampler_state=false
```
An evaluation can be run with:
```
deepqmc task=evaluate hydra.run.dir='runs-12/9_evaluate' task.restdir='runs-12/9/training'
```
Through the CLI all arguments can be changed with the above syntax. You may have to add a + if the argument is not explicitly specified in the default configuration.

As long as you use the basic functionality of DeepQMC you don't need to write your on config files. You can simply overwrite arguments from the comand line. If you want to go beyond small changes and add more advanced configs, it might be convienient to [add your own config files](https://hydra.cc/docs/intro/) at some point. Let me know if you need further help with using hydra. 
A nice side effect of using hydra to start your runs is that now you get a `.hydra` folder in your `training/evaluation` directory that stores all arguments of your run. This might be worth checking out, as it contains the full config with all the arguments you may want to change in your run.
If you want to combine hydra with slurm you can just launch the command from a standard bash script of the slurm job.

Btw. if you want the most accurate results you may want to switch to using [Psiformer](https://arxiv.org/abs/2211.13672) with `ansatz=psiformer` in your hydra command or by loading `config_name='psiformer'` in your `O.py`.

**Comment by Aqsa-sk** [comment]
Hello,
I was successfully able to reset the sampler by the above suggestion. Thank you very much.
Although I have noticed, After resetting the samples the variance overshoots, regardless of increasing or decreasing batch size, is this a common occurrence?

Another quick question regarding the type of sampler:
For my python implementation, I use the following tutorial version:

```
elec_sampler = partial(combine_samplers, samplers=[DecorrSampler(length=20), MetropolisSampler])
sampler_factory = partial(initialize_sampling, elec_sampler=elec_sampler)

```

If I want to switch to Langevin sampler or change the decorrelation length after certain steps by restarting the training, would that be possible?

Thank you for detailed explanations in your response.

**Comment by zenoone** [comment]
Hi,
great to hear that the suggestion worked for you. When resetting the sampler you might see increases variance due to insufficient burn in of the Markov Chains. Although we have an automatic mechanism that detects when the samples are equilibrated, this might not always be sufficient in practice. You can use the `max_eq_steps` and `eq_allow_early_stopping` arguments of the `train` function to disable early stopping of the sampler equilibration and speciy the number of equilibration steps. There are some pathological cases which require special measures, for example systems composed of partially charged subsystems that are separated by a very large distance, but longer equilibration should suffice for most systems. Perhaps you could try that.  Furthermore, if you look at the ewm, which is being accumulated throughout the optimization, the restart resets the average and the respective ewm energies show noise before settling in again.

As for the type of sampler, you are free to choose a different sampler and/or alter the hyperparameters when resetting the training and starting fresh Markov chains:

```
from deepqmc.sampling import LangevinSampler

elec_sampler = partial(combine_samplers, samplers=[DecorrSampler(length=30), LangevinSampler])
sampler_factory = partial(initialize_sampling, elec_sampler=elec_sampler)
```
I'll close this issue. Please feel free to open new issues if you have any other questions.
